lib/os/cbprintf_packaged.c

```diff
@@ -0,0 +1,506 @@
+#define RO_START _image_rom_start
+#define RO_END _image_rom_end
+#elif defined(CONFIG_XTENSA)
+	extern char _rodata_start[];
+	extern char _rodata_end[];
+#define RO_START _rodata_start
+#define RO_END _rodata_end
+#else
+#define RO_START 0
+#define RO_END 0
+#endif
+
+	return ((addr >= (const char *)RO_START) &&
+		(addr < (const char *)RO_END));
+}
+
+/*
+ * va_list creation
+ */
+
+#if defined(__aarch64__)
+/*
+ * Reference:
+ *
+ * Procedure Call Standard for the ARM 64-bit Architecture
+ */
+
+struct __va_list {
+	void	*__stack;
+	void	*__gr_top;
+	void	*__vr_top;
+	int	__gr_offs;
+	int	__vr_offs;
+};
+
+static int cbprintf_via_va_list(cbprintf_cb out, void *ctx,
+				const char *fmt, void *buf)
+{
+	union {
+		va_list ap;
+		struct __va_list __ap;
+	} u;
+
+	/* create a valid va_list with our buffer */
+	u.__ap.__stack = buf;
+	u.__ap.__gr_top = NULL;
+	u.__ap.__vr_top = NULL;
+	u.__ap.__gr_offs = 0;
+	u.__ap.__vr_offs = 0;
+
+	return cbvprintf(out, ctx, fmt, u.ap);
+}
+
+#define VA_STACK_MIN_ALIGN	8
+
+#elif defined(__x86_64__)
+/*
+ * Reference:
+ *
+ * System V Application Binary Interface
+ * AMD64 Architecture Processor Supplement
+ */
+
+struct __va_list {
+	unsigned int gp_offset;
+	unsigned int fp_offset;
+	void *overflow_arg_area;
+	void *reg_save_area;
+};
+
+static int cbprintf_via_va_list(cbprintf_cb out, void *ctx,
+				const char *fmt, void *buf)
+{
+	union {
+		va_list ap;
+		struct __va_list __ap;
+	} u;
+
+	/* create a valid va_list with our buffer */
+	u.__ap.overflow_arg_area = buf;
+	u.__ap.reg_save_area = NULL;
+	u.__ap.gp_offset = (6 * 8);
+	u.__ap.fp_offset = (6 * 8 + 16 * 16);
+
+	return cbvprintf(out, ctx, fmt, u.ap);
+}
+
+#define VA_STACK_MIN_ALIGN	8
+
+#else
+/*
+ * Default implementation shared by many architectures like
+ * 32-bit ARM and Intel.
+ *
+ * We assume va_list is a simple pointer.
+ */
+
+BUILD_ASSERT(sizeof(va_list) == sizeof(void *),
+	     "architecture specific support is needed");
+
+static int cbprintf_via_va_list(cbprintf_cb out, void *ctx,
+				const char *fmt, void *buf)
+{
+	union {
+		va_list ap;
+		void *ptr;
+	} u;
+
+	u.ptr = buf;
+
+	return cbvprintf(out, ctx, fmt, u.ap);
+}
+
+#endif
+
+/*
+ * Special alignment cases
+ */
+
+#if defined(__i386__)
+/* those are not natually aligned on the stack */
+#define VA_STACK_ALIGN_LONGLONG	4
+#define VA_STACK_ALIGN_DOUBLE	4
+#define VA_STACK_ALIGN_INTMAX	4
+#endif
+
+#if defined(__riscv)
+#define VA_STACK_MIN_ALIGN	(__riscv_xlen / 8)
+#endif
+
+/*
+ * Default alignment values if not specified by architecture config
+ */
+
+#ifndef VA_STACK_MIN_ALIGN
+#define VA_STACK_MIN_ALIGN	1
+#endif
+
+#ifndef VA_STACK_ALIGN_LONGLONG
+#define VA_STACK_ALIGN_LONGLONG	__alignof__(long long)
+#endif
+
+#ifndef VA_STACK_ALIGN_DOUBLE
+#define VA_STACK_ALIGN_DOUBLE	__alignof__(double)
+#endif
+
+#ifndef VA_STACK_ALIGN_INTMAX
+#define VA_STACK_ALIGN_INTMAX	__alignof__(intmax_t)
+#endif
+
+
+int cbvprintf_package(void *packaged, size_t *len,
```

Comment:
what about `packaged` buffer alignment. It does not matter when `cbpprintf` is called from the same memory but if package is copied then new buffer must have same alignment as the original one. Maybe require buffer to be aligned?